content/blog/eng/from-0-to-infinite-the-final-keyword.md

```diff
@@ -0,0 +1,333 @@
+
+Java has the same keyword, c# too but it is called "sealed" instead of "final", but more or less have the same behavior.
+To prevent a class from being inherited (in our case we will see only this type of use,for example languages like java support further behavior but in different contexts).
+
+First of all we will try to understand what it is for, then we will try to do some additional reasoning, and finally we will talk about of comments on the net.
+
+The concept that expresses the keyword final is that a class can not be extended, the decision must be taken by the programmer in the design stage utilizing just the final keyword. 
+
+There are situations in which extension by inheritance is a good solution, others where it is not possible and others where it is wrong. The keyword final is a "tool" given to the programmer to express this decision.
+
+So when the consumer of the class will meet the keyword final, it will have to use different solutions from the inheritance. And this decision is imposed from the class designer.
+
+Why someone should set such a strong constraint?. The answers are many, all aim of having a robust,  and more manageable code-base. For example to avoid the proliferation of infinite hierarchies of classes such as VCL (see Visual Component Library), which are difficult to extend and maintain. In this case (even being a well-made library) the use of inheritance has been used a lot.
+
+![big hierarchy](/images/from-0-to-infinite-the-final-keyword/big-hierarchy.png)
+
+When we meet a class with the keyword final, who designed the class not allow us inherit from it, if we want to extend some behavior we must do it by composition or other mechanisms (like events, or plugins). 
+
+I think that this is a good method. In fact, in relation to the time used for the analysis of the design class, through final we provide important information on how she should be useds, simplifying the effort of the developer.
+
+# S.O.L.I.D, Liskov and final
+
+Develop architecture isn't a trivial task, many programmers use inheritance as solution to all problems.  
+
+There are many principles and guidelines to help to develop good design like S.O.L.I.D, design patterns, etc.
+But the problem is that these principles are difficult to apply, they almost always lead to writing more code. 
+Not always this is possible, not always this is accepted. 
+
+Regarding final, in my opinion, one of the principle that best represents it is LSP.
+
+Let's see what **Liskov principle** says in more detail. 
+The canonical statement says the following:
+
+> "you should always be able to substitute the parent class with its derived class without any undesirable behaviour"
+
+> The concept of this principle was introduced by Barbara **Liskov** in a 1987 conference keynote and later published in a paper together with Jannette Wing in 1994. 
```

Comment:
It would be great if you could refer the original paper in the bibliography section

Reply:
https://dl.acm.org/doi/pdf/10.1145/62138.62141
https://pdfs.semanticscholar.org/36be/babeb72287ad9490e1ebab84e7225ad6a9e5.pdf
only for me as a reminder